TESTING.md

```diff
@@ -0,0 +1,96 @@
+
+### Static tests
+
+The static tests suite performs a static analysis on the source code without executing it. It will catch syntax and code style errors.
+
+### Unit tests
+
+The unit tests is a suite of low-level white box tests. They excercise units of code to verify that the different parts work as expected in a fully isolated way. Ideally, these tests should call only public functions, objects and methods, leaving the internals of snapcraft as implementation details that can change without having to modifying any tests. To isolate the units from their environment and dependencies we can replace those with test doubles. In order to set up test doubles, we prefer dependency injection through arguments than excessive mocking. These tests can verify the results checking the output printed to the command line, checking the files created during the excecution, inspecting the calls made to the test doubles and verifying that the expected exceptions were thrown.
+
+These tests are in the `snapcraft/tests` directory.
+
+### Integration tests
+
+The integration tests are a group of suites that excercise snapcraft as a black box. They are only allowed to set up the environment where snapcraft runs and create files; but for the execution phase of the test they can only run the snapcraft command or one of its subcommands. To verify the results they can check the output printed to the command line, the return value of the snapcraft command, and any files created during the execution.
+
+This suite was split in three: plugins, store and other integration tests. This split is artificial, we made it just because the full suite takes more time than what Travis allows for a single job.
+
+These tests are in the `integration_tests` directory, with the `snapcraft.yamls` and other source files for the tests snaps in `integration_tests/snaps`.
+
+### Snaps tests
+
+The snaps tests is a suite of high-level tests that try to simulate real-world scenarios of a user interacting with snapcraft. They cover the call to snapcraft to generate a snap file from the source files of a fully functional project, the installation of the resulting snap, and the execution of the binaries and services of this snap.
+
+These tests are in the `snaps_tests` directory, with the sources for the test snaps in the `demos` directory.
+
+### Setting up the environment
+
+In order to run these tests suites, first you will need to set up your development environment. Follow the steps in the [Hacking guide](HACKING.md) to install for development.
+
+Then, you'll need a few more dependencies:
+
+    sudo apt install squashfs-tools xdelta3 bzr git mercurial subversion
+
+### Running the tests
+
+To run all the tests execute:
+
+    ./runtests.sh
+
+You can selectively run only one of the suites, and apply a filter:
+
+    ./runtests.sh [static|unit|integration|plugins|store|snaps] [pattern]
```

Comment:
Instead of this here, it'd be nicer to have that as a Usage message in the script, and be able to say "check `./runtests.sh --help` for the exact syntax", also because you wouldn't need to search for this documentation later.